Haskell/Haskell.sublime-syntax

```diff
@@ -1,407 +1,1388 @@
+        4: constant.character.escape.hexadecimal.haskell
+        5: constant.character.escape.control.haskell
+
+###[ KEYWORDS AND OPERATORS ]##################################################
+
+  keywords:
+    - match: (?:mdo|do|in|rec|where){{break}}
       scope: keyword.control.context.haskell
-    - match: \b(?:newtype|type)\b
-      scope: keyword.declaration.type.haskell
-    - match: \b(?:data)\b
-      scope: keyword.declaration.data.haskell
-    - match: \b(?:deriving)\b
-      scope: keyword.declaration.data.haskell
-    - match: \b(?:case|of)\b
+    - match: (?:case|of){{break}}
       scope: keyword.control.conditional.select.haskell  # the construct is commonly called "select"
-    - match: \b(?:let|where)\b
+    - match: if{{break}}
+      scope: keyword.control.conditional.if.haskell
+    - match: then{{break}}
+      scope: keyword.control.conditional.then.haskell
+    - match: else{{break}}
+      scope: keyword.control.conditional.else.haskell
+    - match: let{{break}}
       scope: keyword.declaration.variable.haskell
-    - match: \breturn\b
-      scope: keyword.control.flow.return.haskell
-    - match: \b(?:default|otherwise)\b
-      scope: keyword.other.haskell
-    - match: \b(?:(if)|(then)|(else))\b
-      captures:
-        1: keyword.control.conditional.if.haskell
-        2: keyword.control.conditional.then.haskell
-        3: keyword.control.conditional.else.haskell
-
-  operator:
-    - match: (`)[a-zA-Z_'.\d]+(`)
-      # Haskell allows any ordinary function application (elem 4 [1..10])
-      # to be rewritten as an infix expression (4 `elem` [1..10])."
-      scope: keyword.operator.function.infix.haskell
-      captures:
-        1: punctuation.definition.function.begin.haskell
-        2: punctuation.definition.function.end.haskell
-    - match: '\binfix[lr]?\b'
+    - match: proc{{break}}
+      scope: keyword.declaration.function.haskell
+    # 4.4.2 Fixity Declarations
+    - match: infix[lr]?{{break}}
+      scope: keyword.declaration.fixity.haskell
+
+  operators:
+    - include: sequence-separators
+    - include: big-arrow-operators
+    - include: left-arrow-operators
+    - include: right-arrow-operators
+    - include: prefix-parens-operators
+    - include: infix-quoted-operators
+    # Match all not otherwise matched single quotes as promoition operator
+    # Note: Found in real world code but not in specs so far.
+    - match: \'
       scope: keyword.operator.haskell
-    - match: \b(\d+)(?:(\.)(\d+)([eE][-+]?\d+)?|([eE][-+]?\d+))\b
-      scope: meta.number.float.decimal.haskell
-      captures:
-        1: constant.numeric.value.haskell
-        2: punctuation.separator.decimal.haskell
-        3: constant.numeric.value.haskell
-        4: constant.numeric.value.exponent.haskell
-        5: constant.numeric.value.exponent.haskell
-    - match: '{{operator_infix}}'
+    - match: '{{operator_double_colon}}'
+      scope: keyword.operator.double-colon.haskell
```

Comment:
I wouldn't treat the double colon as a normal operator either, more like a `keyword.declaration`.

Reply:
I don't want to introduce another ambiguity for word vs. symbol like keywords and therefor also don't like that scope for lambda arrows to be honest.

Reply:
Have we had an RFC for this?

Reply:
There were several discussions about word vs. symbol operators in pending PRs for instance.

Disuccsion about lambda function declarations was at  https://github.com/sublimehq/Packages/pull/2617

Didn't recall the conclusion myself, hence should probably move it to an RFC to not loose it (again).

---

Erlang also has this concept of double-colon separated types, which uses `punctuation.separator` atm.

Example:

```Erlang
-spec formaterror(Tag :: term()) -> string().
```

It is just a separator between function parameters and their type definitions. So `punctuation.separator` would be an alternative with regards to how Erlang handles that construct.

Reply:
`punctuation` is definitely better than `keyword.operator` and would certainly be the correct choice inside record type definitions.

I don't think the Erlang example fits, however, because in Haskell the `::` are integral to the type of statement on this line (not in a record type) and denote a signature declaration. A third opinion should be helpful.